From 00655b931e8f5e4a6b59665c3e7d7adeddc20adf Mon Sep 17 00:00:00 2001
From: Haibo Chen <haibo.chen@nxp.com>
Date: Wed, 21 Mar 2018 18:26:46 +0800
Subject: [PATCH 3520/5242] MLK-17883 mmc: add delay after CMD6 before sending
 CMD13

commit  57d57e76f938f2814bf5d4affd3fff4adec03d2a from
https://source.codeaurora.org/external/imx/linux-imx.git

Android report data timeout issue on the Micron eMMC of i.mx8mscale
evk board. Before using CMD25 to write multi block data, driver will
use CMD6 to flush cache. But some CMD25 will get both data timeout
interrupt and transfer complete interrupt. If add delay after busy
check, and before sending CMD13, this issue gone.

We meet similar issue on Sandisk eMMC before, and already have a
commit 581925fe325a ("MLK-11685-1 mmc: add delay after CMD6 befoer
sending CMD13 for sandisk"), but when cherry-pick to 4.9 branch,
meet conflicts, put the delay in the wrong place. So meet the similar
issue again.

This patch fix this, move the dealy in the right place, right before
CMD13.

Reviewed-by: Dong Aisheng <Aisheng.dong@nxp.com>
Signed-off-by: Haibo Chen <haibo.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mmc/core/mmc_ops.c |   12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/mmc/core/mmc_ops.c b/drivers/mmc/core/mmc_ops.c
index 94ddf18..d933cc6e 100644
--- a/drivers/mmc/core/mmc_ops.c
+++ b/drivers/mmc/core/mmc_ops.c
@@ -570,12 +570,6 @@ int __mmc_switch(struct mmc_card *card, u8 set, u8 index, u8 value,
 	if (!use_busy_signal)
 		goto out;
 
-	/*
-	 * WORKAROUND: for Sandisk eMMC cards, it might need certain delay
-	 * before sending CMD13 after CMD6
-	 */
-	mdelay(1);
-
 	/*If SPI or used HW busy detection above, then we don't need to poll. */
 	if (((host->caps & MMC_CAP_WAIT_WHILE_BUSY) && use_r1b_resp) ||
 		mmc_host_is_spi(host))
@@ -591,6 +585,12 @@ int __mmc_switch(struct mmc_card *card, u8 set, u8 index, u8 value,
 	if (timing)
 		mmc_set_timing(host, timing);
 
+	/*
+	 * WORKAROUND: for Sandisk eMMC cards, it might need certain delay
+	 * before sending CMD13 after CMD6
+	 */
+	mdelay(1);
+
 	if (send_status) {
 		err = mmc_switch_status(card);
 		if (err && timing)
-- 
1.7.9.5

